lib/assets/javascripts/set_builder.js

```diff
@@ -275,22 +275,49 @@ SetBuilder.Set = function(_raw_data) {
   if(!_raw_data) _raw_data = [];
   
   var _constraints = [];
-  _raw_data.__each(function(line) {
-    var trait_name = line[0];
-    var negative = false;
-    if(trait_name[0] == '!') {
-      negative = true;
-      trait_name = trait_name.slice(1);
-    }
-    var trait = SetBuilder.traits().__find(trait_name);
-    var args = line.slice(1);
-    if(trait) {
-      _constraints.push(trait.apply(args).negate(negative));
-    } else if(window.console && window.console.log) {
-      window.console.log('trait not found with name "' + line[0] + '"');
-    }
-  });
   
+  if(_raw_data.constructor === Array) {
+    _raw_data.__each(function(line) {
+      var trait_name = line[0];
+      var negative = false;
+    
+      if(trait_name[0] == '!') {
+        negative = true;
+        trait_name = trait_name.slice(1);
+      }
+      var trait = SetBuilder.traits().__find(trait_name);
+      var args = line.slice(1);
+      if(trait) {
+        _constraints.push(trait.apply(args).negate(negative));
+      } else if(window.console && window.console.log) {
+        window.console.log('trait not found with name "' + line[0] + '"');
+      }
+    });
+  } else if (typeof _raw_data == 'object') {
+    Object.keys(_raw_data).__each(function(key) {
+      var traitHash = _raw_data[key];
+      var trait_name = traitHash.trait;
+      var negative = !!traitHash.negative;
+      var trait = SetBuilder.traits().__find(trait_name);
+      var args = [];
+      if(traitHash.direct_object) args.push(traitHash.direct_object);
+      
+      if(traitHash.modifiers) {
+        Object.keys(traitHash.modifiers).__each(function(key) {
+          var modifierHash = traitHash.modifiers[key];
+          var convertedModifier = {};
+          convertedModifier[modifierHash.operator] = modifierHash.values ? Object.values(modifierHash.values) : [];
+          args.push(convertedModifier);
+        });
+      }
+      
+      if(trait) {
+        _constraints.push(trait.apply(args).negate(negative));
+      } else if(window.console && window.console.log) {
+        window.console.log('trait not found with name "' + trait_name + '"');
+      }
+    });
+  }
```

Comment:
Why'd you decide to have two entirely separate paths through this method?

Would it not work to have the constructor convert a hash-style set into an array-style set (and then sequester the knowledge about _reading_ a set in a single place)?